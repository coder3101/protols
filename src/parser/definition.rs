use async_lsp::lsp_types::{Location, Position, Range};
use tracing::info;

use crate::{parser::nodekind::NodeKind, utils::ts_to_lsp_position};

use super::ParsedTree;

impl ParsedTree {
    pub fn definition(&self, pos: &Position, content: impl AsRef<[u8]>) -> Vec<Location> {
        let text = self.get_node_text_at_position(pos, content.as_ref());
        info!("Looking for definition of: {:?}", text);

        match text {
            Some(text) => self
                .filter_nodes(NodeKind::is_userdefined)
                .into_iter()
                .filter(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error") == text)
                .map(|n| Location {
                    uri: self.uri.clone(),
                    range: Range {
                        start: ts_to_lsp_position(&n.start_position()),
                        end: ts_to_lsp_position(&n.end_position()),
                    },
                })
                .collect(),
            None => vec![],
        }
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::{Position, Url};
    use insta::assert_yaml_snapshot;

    use crate::parser::ProtoParser;

    #[test]
    fn test_goto_definition() {
        let url: Url = "file://foo/bar.proto".parse().unwrap();
        let posinvalid = Position {
            line: 0,
            character: 1,
        };
        let posauthor = Position {
            line: 10,
            character: 5,
        };
        let contents = include_str!("input/test_goto_definition.proto");
        let parsed = ProtoParser::new().parse(url, contents);

        assert!(parsed.is_some());
        let tree = parsed.unwrap();
        assert_yaml_snapshot!(tree.definition(&posauthor, contents));
        assert_yaml_snapshot!(tree.definition(&posinvalid, contents));
    }
}
