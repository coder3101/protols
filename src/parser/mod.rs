use async_lsp::lsp_types::Url;
use tree_sitter::Tree;

mod definition;
mod diagnostics;
mod docsymbol;
mod hover;
mod rename;
mod tree;

pub struct ProtoParser {
    parser: tree_sitter::Parser,
}

pub struct ParsedTree {
    pub uri: Url,
    tree: Tree,
}

impl ProtoParser {
    pub fn new() -> Self {
        let mut parser = tree_sitter::Parser::new();
        if let Err(e) = parser.set_language(&protols_tree_sitter_proto::language()) {
            panic!("failed to set ts language parser {:?}", e);
        }
        Self { parser }
    }

    pub fn parse(&mut self, uri: Url, contents: impl AsRef<[u8]>) -> Option<ParsedTree> {
        self.parser
            .parse(contents, None)
            .map(|t| ParsedTree { tree: t, uri })
    }
}
