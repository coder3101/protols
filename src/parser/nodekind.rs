use async_lsp::lsp_types::SymbolKind;
use tree_sitter::Node;

#[allow(unused)]
pub enum NodeKind {
    Identifier,
    Error,
    MessageName,
    EnumName,
    FieldName,
    ServiceName,
    RpcName,
    PackageName,
}

#[allow(unused)]
impl NodeKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            NodeKind::Identifier => "identifier",
            NodeKind::Error => "ERROR",
            NodeKind::MessageName => "message_name",
            NodeKind::EnumName => "enum_name",
            NodeKind::FieldName => "message_or_enum_type",
            NodeKind::ServiceName => "service_name",
            NodeKind::RpcName => "rpc_name",
            NodeKind::PackageName => "package_name",
        }
    }

    pub fn is_identifier(n: &Node) -> bool {
        n.kind() == "identifier"
    }

    pub fn is_error(n: &Node) -> bool {
        n.kind() == "ERROR"
    }

    pub fn is_userdefined(n: &Node) -> bool {
        matches!(n.kind(), "message_name" | "enum_name")
    }

    pub fn is_actionable(n: &Node) -> bool {
        matches!(
            n.kind(),
            "message_name" | "enum_name" | "message_or_enum_type" | "rpc_name" | "service_name"
        )
    }

    pub fn to_symbolkind(n: &Node) -> SymbolKind {
        match n.kind() {
            "message_name" => SymbolKind::STRUCT,
            "enum_name" => SymbolKind::ENUM,
            _ => SymbolKind::NULL,
        }
    }
}


