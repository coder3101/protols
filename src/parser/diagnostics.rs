use async_lsp::lsp_types::{Diagnostic, DiagnosticSeverity, PublishDiagnosticsParams, Range, Url};

use crate::utils::ts_to_lsp_position;

use super::{nodekind::NodeKind, ParsedTree};

impl ParsedTree {
    pub fn collect_parse_errors(&self, uri: &Url) -> PublishDiagnosticsParams {
        let diagnostics = self
            .filter_node(NodeKind::is_error)
            .into_iter()
            .map(|n| Diagnostic {
                range: Range {
                    start: ts_to_lsp_position(&n.start_position()),
                    end: ts_to_lsp_position(&n.end_position()),
                },
                severity: Some(DiagnosticSeverity::ERROR),
                source: Some("protols".to_string()),
                message: "Syntax error".to_string(),
                ..Default::default()
            })
            .collect();
        PublishDiagnosticsParams {
            uri: uri.clone(),
            diagnostics,
            version: None,
        }
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::{DiagnosticSeverity, Position, Range, Url};

    use crate::parser::ProtoParser;

    #[test]
    fn test_collect_parse_error() {
        let url = "file://foo/bar.proto";
        let contents = r#"syntax = "proto3";

package test;

message Foo {
	reserved 1;
	reserved "baz";
	int bar = 2;
}"#;

        let parsed = ProtoParser::new().parse(contents);
        assert!(parsed.is_some());
        let tree = parsed.unwrap();
        let diagnostics = tree.collect_parse_errors(&url.parse().unwrap());
        assert_eq!(diagnostics.uri, Url::parse(url).unwrap());
        assert_eq!(diagnostics.diagnostics.len(), 0);

        let url = "file://foo/bar.proto";
        let contents = r#"syntax = "proto3";

package com.book;

message Book {
    message Author {
        string name;
        string country = 2;
    };
}"#;
        let parsed = ProtoParser::new().parse(contents);
        assert!(parsed.is_some());
        let tree = parsed.unwrap();
        let diagnostics = tree.collect_parse_errors(&url.parse().unwrap());

        assert_eq!(diagnostics.uri, Url::parse(url).unwrap());
        assert_eq!(diagnostics.diagnostics.len(), 1);

        let error = &diagnostics.diagnostics[0];
        assert_eq!(error.severity, Some(DiagnosticSeverity::ERROR));
        assert_eq!(error.source, Some("protols".to_owned()));
        assert_eq!(error.message, "Syntax error");
        assert_eq!(
            error.range,
            Range {
                start: Position {
                    line: 6,
                    character: 8
                },
                end: Position {
                    line: 6,
                    character: 19
                }
            }
        );
    }
}
