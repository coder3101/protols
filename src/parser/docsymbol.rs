use async_lsp::lsp_types::{DocumentSymbol, Range};
use tree_sitter::TreeCursor;

use crate::utils::ts_to_lsp_position;

use super::{ nodekind::NodeKind, ParsedTree};

#[derive(Default)]
pub(super) struct DocumentSymbolTreeBuilder {
    // The stack are things we're still in the process of building/parsing.
    stack: Vec<(usize, DocumentSymbol)>,
    // The found are things we've finished processing/parsing, at the top level of the stack.
    found: Vec<DocumentSymbol>,
}

impl DocumentSymbolTreeBuilder {
    pub(super) fn push(&mut self, node: usize, symbol: DocumentSymbol) {
        self.stack.push((node, symbol));
    }

    pub(super) fn maybe_pop(&mut self, node: usize) {
        let should_pop = self.stack.last().map_or(false, |(n, _)| *n == node);
        if should_pop {
            let (_, explored) = self.stack.pop().unwrap();
            if let Some((_, parent)) = self.stack.last_mut() {
                parent.children.as_mut().unwrap().push(explored);
            } else {
                self.found.push(explored);
            }
        }
    }

    pub(super) fn build(self) -> Vec<DocumentSymbol> {
        self.found
    }
}

impl ParsedTree {
    pub fn find_document_locations(&self, content: impl AsRef<[u8]>) -> Vec<DocumentSymbol> {
        let mut builder = DocumentSymbolTreeBuilder::default();
        let content = content.as_ref();

        let mut cursor = self.tree.root_node().walk();
        self.find_document_locations_inner(&mut builder, &mut cursor, content);

        builder.build()
    }

    fn find_document_locations_inner(
        &self,
        builder: &mut DocumentSymbolTreeBuilder,
        cursor: &'_ mut TreeCursor,
        content: &[u8],
    ) {
        loop {
            let node = cursor.node();

            if NodeKind::is_userdefined(&node) {
                let name = node.utf8_text(content).unwrap();
                let kind = NodeKind::to_symbolkind(&node);
                let detail = self.find_preceding_comments(node.id(), content);
                let message = node.parent().unwrap();

                // https://github.com/rust-lang/rust/issues/102777
                #[allow(deprecated)]
                let new_symbol = DocumentSymbol {
                    name: name.to_string(),
                    detail,
                    kind,
                    tags: None,
                    deprecated: None,
                    range: Range {
                        start: ts_to_lsp_position(&message.start_position()),
                        end: ts_to_lsp_position(&message.end_position()),
                    },
                    selection_range: Range {
                        start: ts_to_lsp_position(&node.start_position()),
                        end: ts_to_lsp_position(&node.end_position()),
                    },
                    children: Some(vec![]),
                };

                builder.push(message.id(), new_symbol);
            }

            if cursor.goto_first_child() {
                self.find_document_locations_inner(builder, cursor, content);
                builder.maybe_pop(node.id());
                cursor.goto_parent();
            }

            if !cursor.goto_next_sibling() {
                break;
            }
        }
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::{DocumentSymbol, Position, Range, SymbolKind};

    use crate::parser::ProtoParser;

    #[test]
    #[allow(deprecated)]
    fn test_document_symbols() {
        let contents = r#"syntax = "proto3";

package com.symbols;

// outer 1 comment
message Outer1 {
    message Inner1 {
        string name = 1;
    };

    Inner1 i = 1;
}

message Outer2 {
    message Inner2 {
        string name = 1;
    };
    // Inner 3 comment here
    message Inner3 {
        string name = 1;

        enum X {
            a = 1;
            b = 2;
        }
    }
    Inner1 i = 1;
    Inner2 y = 2;
}

"#;
        let parsed = ProtoParser::new().parse(contents);
        assert!(parsed.is_some());
        let tree = parsed.unwrap();
        let res = tree.find_document_locations(contents);

        assert_eq!(res.len(), 2);
        assert_eq!(
            res,
            vec!(
                DocumentSymbol {
                    name: "Outer1".to_string(),
                    detail: Some("outer 1 comment".to_string()),
                    kind: SymbolKind::STRUCT,
                    tags: None,
                    range: Range {
                        start: Position::new(5, 0),
                        end: Position::new(11, 1),
                    },
                    selection_range: Range {
                        start: Position::new(5, 8),
                        end: Position::new(5, 14),
                    },
                    children: Some(vec!(DocumentSymbol {
                        name: "Inner1".to_string(),
                        detail: None,
                        kind: SymbolKind::STRUCT,
                        tags: None,
                        deprecated: None,
                        range: Range {
                            start: Position::new(6, 4),
                            end: Position::new(8, 5),
                        },
                        selection_range: Range {
                            start: Position::new(6, 12),
                            end: Position::new(6, 18),
                        },
                        children: Some(vec!()),
                    },)),
                    deprecated: None,
                },
                DocumentSymbol {
                    name: "Outer2".to_string(),
                    detail: None,
                    kind: SymbolKind::STRUCT,
                    tags: None,
                    range: Range {
                        start: Position::new(13, 0),
                        end: Position::new(28, 1),
                    },
                    selection_range: Range {
                        start: Position::new(13, 8),
                        end: Position::new(13, 14),
                    },
                    children: Some(vec!(
                        DocumentSymbol {
                            name: "Inner2".to_string(),
                            detail: None,
                            kind: SymbolKind::STRUCT,
                            tags: None,
                            deprecated: None,
                            range: Range {
                                start: Position::new(14, 4),
                                end: Position::new(16, 5),
                            },
                            selection_range: Range {
                                start: Position::new(14, 12),
                                end: Position::new(14, 18),
                            },
                            children: Some(vec!()),
                        },
                        DocumentSymbol {
                            name: "Inner3".to_string(),
                            detail: Some("Inner 3 comment here".to_string()),
                            kind: SymbolKind::STRUCT,
                            tags: None,
                            deprecated: None,
                            range: Range {
                                start: Position::new(18, 4),
                                end: Position::new(25, 5),
                            },
                            selection_range: Range {
                                start: Position::new(18, 12),
                                end: Position::new(18, 18),
                            },
                            children: Some(vec!(DocumentSymbol {
                                name: "X".to_string(),
                                detail: None,
                                kind: SymbolKind::ENUM,
                                tags: None,
                                deprecated: None,
                                range: Range {
                                    start: Position::new(21, 8),
                                    end: Position::new(24, 9),
                                },
                                selection_range: Range {
                                    start: Position::new(21, 13),
                                    end: Position::new(21, 14),
                                },
                                children: Some(vec!()),
                            })),
                        }
                    )),
                    deprecated: None,
                },
            )
        );
    }
}
