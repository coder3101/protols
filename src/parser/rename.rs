use std::collections::HashMap;

use async_lsp::lsp_types::{Position, Range, TextEdit, WorkspaceEdit};

use crate::utils::ts_to_lsp_position;

use super::{nodekind::NodeKind, ParsedTree};

impl ParsedTree {
    pub fn can_rename(&self, pos: &Position) -> Option<Range> {
        self.get_node_at_position(pos)
            .filter(NodeKind::is_identifier)
            .map(|n| n.parent().unwrap()) // Safety: Identifier must have a parent node
            .filter(NodeKind::is_actionable)
            .map(|n| Range {
                start: ts_to_lsp_position(&n.start_position()),
                end: ts_to_lsp_position(&n.end_position()),
            })
    }

    pub fn rename(
        &self,
        pos: &Position,
        new_text: &str,
        content: impl AsRef<[u8]>,
    ) -> Option<WorkspaceEdit> {
        let old_text = self
            .get_node_text_at_position(pos, content.as_ref())
            .unwrap_or_default();

        let mut changes = HashMap::new();

        let diff: Vec<_> = self
            .filter_nodes(NodeKind::is_identifier)
            .into_iter()
            .filter(|n| n.utf8_text(content.as_ref()).unwrap() == old_text)
            .map(|n| TextEdit {
                new_text: new_text.to_string(),
                range: Range {
                    start: ts_to_lsp_position(&n.start_position()),
                    end: ts_to_lsp_position(&n.end_position()),
                },
            })
            .collect();

        if diff.is_empty() {
            return None;
        }

        changes.insert(self.uri.clone(), diff);

        Some(WorkspaceEdit {
            changes: Some(changes),
            ..Default::default()
        })
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::{Position, Url};
    use insta::assert_yaml_snapshot;

    use crate::parser::ProtoParser;

    #[test]
    fn test_rename() {
        let uri: Url = "file://foo/bar.proto".parse().unwrap();
        let pos_book_rename = Position {
            line: 5,
            character: 9,
        };
        let pos_author_rename = Position {
            line: 21,
            character: 10,
        };
        let pos_non_renamble = Position {
            line: 24,
            character: 4,
        };
        let contents = include_str!("input/test_rename.proto");

        let parsed = ProtoParser::new().parse(uri.clone(), contents);
        assert!(parsed.is_some());
        let tree = parsed.unwrap();

        assert_yaml_snapshot!(tree.rename(&pos_book_rename, "Kitab", contents));
        assert_yaml_snapshot!(tree.rename(&pos_author_rename, "Writer", contents));
        assert_yaml_snapshot!(tree.rename(&pos_non_renamble, "Doesn't matter", contents));
    }

    #[test]
    fn test_can_rename() {
        let uri: Url = "file://foo/bar/test.proto".parse().unwrap();
        let pos_rename = Position {
            line: 5,
            character: 9,
        };
        let pos_non_rename = Position {
            line: 2,
            character: 2,
        };
        let contents = include_str!("input/test_can_rename.proto");
        let parsed = ProtoParser::new().parse(uri.clone(), contents);
        assert!(parsed.is_some());

        let tree = parsed.unwrap();
        assert_yaml_snapshot!(tree.can_rename(&pos_rename));
        assert_yaml_snapshot!(tree.can_rename(&pos_non_rename));
    }
}
