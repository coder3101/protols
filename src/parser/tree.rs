use async_lsp::lsp_types::Position;
use tree_sitter::{Node, TreeCursor};

use crate::utils::lsp_to_ts_point;

use super::{nodekind::NodeKind, ParsedTree};

impl ParsedTree {
    pub(super) fn walk_and_collect_filter<'a>(
        cursor: &mut TreeCursor<'a>,
        f: fn(&Node) -> bool,
    ) -> Vec<Node<'a>> {
        let mut v = vec![];

        loop {
            let node = cursor.node();

            if f(&node) {
                v.push(node)
            }

            if cursor.goto_first_child() {
                v.extend(Self::walk_and_collect_filter(cursor, f));
                cursor.goto_parent();
            }

            if !cursor.goto_next_sibling() {
                break;
            }
        }

        v
    }

    pub(super) fn advance_cursor_to(cursor: &mut TreeCursor<'_>, nid: usize) -> bool {
        loop {
            let node = cursor.node();
            if node.id() == nid {
                return true;
            }
            if cursor.goto_first_child() {
                if Self::advance_cursor_to(cursor, nid) {
                    return true;
                }
                cursor.goto_parent();
            }
            if !cursor.goto_next_sibling() {
                return false;
            }
        }
    }

    pub(super) fn get_node_text_at_position<'a>(
        &'a self,
        pos: &Position,
        content: &'a [u8],
    ) -> Option<&'a str> {
        self.get_node_at_position(pos)
            .map(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error"))
    }

    pub(super) fn get_actionable_node_text_at_position<'a>(
        &'a self,
        pos: &Position,
        content: &'a [u8],
    ) -> Option<&'a str> {
        self.get_actionable_node_at_position(pos)
            .map(|n| n.utf8_text(content.as_ref()).expect("utf-8 parse error"))
    }

    pub(super) fn get_actionable_node_at_position<'a>(
        &'a self,
        pos: &Position,
    ) -> Option<Node<'a>> {
        self.get_node_at_position(pos)
            .map(|n| {
                if NodeKind::is_actionable(&n) {
                    n
                } else {
                    n.parent().unwrap()
                }
            })
            .filter(NodeKind::is_actionable)
    }

    pub(super) fn get_node_at_position<'a>(&'a self, pos: &Position) -> Option<Node<'a>> {
        let pos = lsp_to_ts_point(pos);
        self.tree.root_node().descendant_for_point_range(pos, pos)
    }

    pub(super) fn filter_node(&self, f: fn(&Node) -> bool) -> Vec<Node> {
        let mut cursor = self.tree.root_node().walk();
        Self::walk_and_collect_filter(&mut cursor, f)
    }
}

#[cfg(test)]
mod test {
    use tree_sitter::Node;

    use crate::parser::ProtoParser;

    fn is_message(n: &Node) -> bool {
        n.kind() == "message_name"
    }

    #[test]
    fn test_find_children_by_kind() {
        let contents = r#"syntax = "proto3";

package com.book;

message Book {

    message Author {
        string name = 1;
        string country = 2;
    };
    // This is a multi line comment on the field name
    // Of a message called Book
    int64 isbn = 1;
    string title = 2;
    Author author = 3;
}
"#;
        let parsed = ProtoParser::new().parse(contents);
        assert!(parsed.is_some());
        let tree = parsed.unwrap();
        let nodes = tree.filter_node(is_message);

        assert_eq!(nodes.len(), 2);

        let names: Vec<_> = nodes
            .into_iter()
            .map(|n| n.utf8_text(contents.as_ref()).unwrap())
            .collect();
        assert_eq!(names[0], "Book");
        assert_eq!(names[1], "Author");
    }
}
