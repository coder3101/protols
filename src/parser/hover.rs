use async_lsp::lsp_types::MarkedString;
use tree_sitter::Node;

use crate::nodekind::NodeKind;

use super::ParsedTree;

impl ParsedTree {
    pub(super) fn find_preceding_comments(
        &self,
        nid: usize,
        content: impl AsRef<[u8]>,
    ) -> Option<String> {
        let root = self.tree.root_node();
        let mut cursor = root.walk();

        Self::advance_cursor_to(&mut cursor, nid);
        if !cursor.goto_parent() {
            return None;
        }

        if !cursor.goto_previous_sibling() {
            return None;
        }

        let mut comments = vec![];
        while cursor.node().kind() == "comment" {
            let node = cursor.node();
            let text = node
                .utf8_text(content.as_ref())
                .expect("utf-8 parser error")
                .trim()
                .trim_start_matches("//")
                .trim();

            comments.push(text);

            if !cursor.goto_previous_sibling() {
                break;
            }
        }
        if !comments.is_empty() {
            comments.reverse();
            Some(comments.join("\n"))
        } else {
            None
        }
    }

    pub fn hover(&self, identifier: &str, content: impl AsRef<[u8]>) -> Vec<String> {
        let mut results = vec![];
        self.hover_impl(identifier, self.tree.root_node(), &mut results, content);
        results
    }

    fn hover_impl(
        &self,
        identifier: &str,
        n: Node,
        v: &mut Vec<String>,
        content: impl AsRef<[u8]>,
    ) {
        if identifier.is_empty() {
            return;
        }

        match identifier.split_once('.') {
            Some((parent, child)) => {
                let child_node = self
                    .filter_nodes_from(n, NodeKind::is_userdefined)
                    .into_iter()
                    .find(|n| n.utf8_text(content.as_ref()).expect("utf8-parse error") == parent)
                    .and_then(|n| n.parent());

                if let Some(inner) = child_node {
                    self.hover_impl(child, inner, v, content);
                }
            }
            None => {
                let comments: Vec<String> = self
                    .filter_nodes_from(n, NodeKind::is_userdefined)
                    .into_iter()
                    .filter(|n| {
                        n.utf8_text(content.as_ref()).expect("utf-8 parse error") == identifier
                    })
                    .filter_map(|n| self.find_preceding_comments(n.id(), content.as_ref()))
                    .collect();

                v.extend(comments);
            }
        }
    }
}

#[cfg(test)]
mod test {
    use async_lsp::lsp_types::Url;
    use insta::assert_yaml_snapshot;

    use crate::parser::ProtoParser;

    #[test]
    fn test_hover() {
        let uri: Url = "file://foo.bar/p.proto".parse().unwrap();
        let contents = include_str!("input/test_hover.proto");
        let parsed = ProtoParser::new().parse(uri.clone(), contents);

        assert!(parsed.is_some());
        let tree = parsed.unwrap();

        let res = tree.hover("Book", contents);
        assert_yaml_snapshot!(res);

        let res = tree.hover("", contents);
        assert_yaml_snapshot!(res);

        let res = tree.hover("Book.Author", contents);
        assert_yaml_snapshot!(res);

        let res = tree.hover("Comic.Author", contents);
        assert_yaml_snapshot!(res);

        let res = tree.hover("Author", contents);
        assert_yaml_snapshot!(res);
    }
}
