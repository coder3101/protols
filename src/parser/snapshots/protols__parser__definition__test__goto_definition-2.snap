---
source: src/parser/definition.rs
expression: "tree.definition(&posinvalid, contents)"
---
[]
